File: KARATSUBA_Algorithm.py
def getLength(value):
    count = 0
    while (value != 0):
        count = count + 1
        value = value // 10
    return count


def karatsuba(x, y):
    if (getLength(x) == 1 or getLength(y) == 1):
        return x * y
    else:
        n = max(getLength(x), getLength(y))
        nby2 = n / 2
        a = x // 10 ** (nby2)
        b = x % 10 ** (nby2)
        c = y // 10 ** (nby2)
        d = y % 10 ** (nby2)
        ac = karatsuba(a, c)
        bd = karatsuba(b, d)
        ad_plus_bc = karatsuba(a + b, c + d) - ac - bd
        prod = ac * 10 ** (2 * nby2) + (ad_plus_bc * 10 ** nby2) + bd
        return prod


a = int(input("Enter the first number :- "))
b = int(input("Enter the second number :- "))

print(karatsuba(a, b))

File: KARATSUBA ALGORITHM.cpp
#include <iostream>
#include <math.h>

int getLength(long long value)
{
    int counter = 0;

    while(value != 0)
    {
        counter++;
        value = value / 10;
    }
    return counter;
}

long long power(int base, int exponent)
{
    long long result = 1;
    while(exponent != 0)
    {
        result = result*base;
        exponent--;
    }
    return result;
}

long long multiply(long long a, long long b)
{
    if(getLength(a) == 1 || getLength(b) == 1)
    {
        return a*b;
    }

    else
    {
        int n = fmax(getLength(a), getLength(b));
        int nby2 = n/2;

        long long aL = (a / power(10, nby2));
        long long aR = (a % power(10, nby2));

        long long bL = (b / power(10, nby2));
        long long bR = (b % power(10, nby2));

        long long x1 = multiply(aL, bL);
        long long x2 = multiply((aL+aR), (bL+bR));
        long long x3 = multiply(aR, bR);

        return (x1*power(10, n) + (x2 - x1 - x3)*power(10, nby2) + x3);
    }
}

int main()
{
    long long int a, b;

    std::cout << "Enter the first number :- ";
    std::cin >> a;
    std::cout << "Enter the second number :- ";
    std::cin >> b;

    std::cout << "The multiplication result :- " << multiply(a,b) << std::endl;
}
